### DISCUSSION TITLE
Implicit callbacks/lambdas

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [vlang/v#11766] [Ideas]
I haven't really spotted this feature, so I thought I'd bring it up here.

Quite a while ago, a similiar feature was introduced into AngelScript where passing a callback with no type definitions would have the interpreter assume those types according to the parameter of the function the callback was passed into.

Especially with projects like vex, something akin to this would be quite halpful:

```v
fn do_something(trigger string, callback fn(int, int)int) {...}

do_something("foo", (a, b) { return a + b })
```

Alternatively, you could borrow from Objective-C, where a caret (`^`) is used to indicate a block, which is *basically* a callback as well. This would make the above look like this: `do_something("foo", ^(a, b) { return a + b })`.

This is also useful for sorting through maps and chaining actions together. A stupid, but effective way of utilizing callbacks - or, *predicates*  in this scenario - can be seen here: https://gist.github.com/IngwiePhoenix/501942963259da21def4a6fe0cc1f36a#gistcomment-3189420

What do you think? The current way is to completely type out everything, when only so little is really and actually needed - since the types could be deduced from the callee.

**Comment by olup** [comment]
You can use callback in V already by passing around anonymous functions, exactly as you did in your example. The only thing missing to have a completely practical setup is closure support (inherit values from the parent function context), which is in syntax discussion.

**Comment by IngwiePhoenix** [comment]
Yes - but you *have* to type out the entire declaration:

```
$ v run dork.v
dork.v:8:13: error: undefined ident: `s`
    6 | fn main() {
    7 |   trigger("meep", fn(s) {
    8 |     println(s)
      |             ^
    9 |   })
   10 | }
dork.v:7:19: error: cannot use `fn (s)` as `fn (string)` in argument 2 to `trigger`
    5 |
    6 | fn main() {
    7 |   trigger("meep", fn(s) {
      |                   ~~~~~~~
    8 |     println(s)
    9 |   })
```

The full example:
```v
fn trigger(s string, cb fn(string)) {
  println("Triggering!")
  cb(s)
}

fn main() {
  trigger("meep", fn(s) {
    println(s)
  })
}
```

For small callbacks, typing out the entire typing might actually be overwhelming. Imagine the following:

```v
a := [1, 2, 3]
a.each(fn(i){ return i+1 })
```

Now, in this, you`d have to specify `fn(i int)` instead. This would grow longer in projects like Vex:

```v
app.get("/", fn(req, res) {
  // ...
})
```

Here, `req` and `res` both need to have their types added: `(req vex.Request, res mut vex.Response)`. Not terrible - but for what this and other projects want to achieve, quite a lot.

**Comment by olup** [comment]
I see. I believe you can also declare a type to capture the entire callback
declaration, making it as small as a single word for subsequent usage.

Le dim. 3 janv. 2021 à 16:51, Ingwie Phoenix <notifications@github.com> a
écrit :

**Comment by olup** [comment]
But I agree that anonymous function type inference should be nice !

Le dim. 3 janv. 2021 à 17:09, Loup Topalian <loup.topalian@gmail.com> a
écrit :

**Comment by JalonSolov** [comment]
Typing it all out isn't a bad thing.  It is explicit, and there is no confusion about what is happening or why.

**Comment by ntrel** [comment]
> Typing it all out isn't a bad thing. It is explicit, and there is no confusion

That is overkill when the function body is only returning one expression. V can't really say it supports functional programming well if it doesn't support lambdas. Lambdas don't have statements, only an expression, and need at least return type inference: `odd := fn(int i){i & 1 == 1}`. This is *trivial* to support. The programmer can still write `fn(int i) bool {return i & 1 == 1}` (if they get paid by the character) - that option should only be used for complex expressions where the type is not clear.

V **forces** type inference for variables. Not supporting it for lambda return types is just weird.

**Comment by JalonSolov** [comment]
Where does it say V supports functional programming?

**Comment by ntrel** [comment]
> Where does it say V supports functional programming?

It supports `map` and `filter` methods, which are functional. V should be able to say it supports functional programming well, it just needs to support lambdas properly.

**Comment by IngwiePhoenix** [comment]
> Typing it all out isn't a bad thing. It is explicit, and there is no confusion about what is happening or why.

You say that, and yet there is compiler magic in `or` blocks with `err` and `errcode`. Also for `.map()` and `.filter()`. V does a *lot* of implicit stuff - so why not callbacks too?
